src/coreclr/jit/gentree.cpp

```diff
@@ -18541,12 +18541,24 @@ CORINFO_CLASS_HANDLE Compiler::gtGetClassHandle(GenTree* tree, bool* pIsExact, b
                 if ((fldSeq != nullptr) && (fldSeq->GetOffset() == base->AsIntCon()->IconValue()))
                 {
                     CORINFO_FIELD_HANDLE fldHandle = base->AsIntCon()->gtFieldSeq->GetFieldHandle();
-                    objClass                       = gtGetFieldClassHandle(fldHandle, pIsExact, pIsNonNull);
+                    objClass = gtGetFieldClassHandle(fldHandle, NO_CLASS_HANDLE, pIsExact, pIsNonNull);
                 }
             }
             else if (base->OperIs(GT_FIELD_ADDR))
             {
-                objClass = gtGetFieldClassHandle(base->AsFieldAddr()->gtFldHnd, pIsExact, pIsNonNull);
+                CORINFO_FIELD_HANDLE fldHandle   = base->AsFieldAddr()->gtFldHnd;
+                CORINFO_CLASS_HANDLE fldOwnerCls = NO_CLASS_HANDLE;
+
+                if (base->gtGetOp1() != nullptr)
+                {
+                    // If it's an instance field, try to grab the class handle from the instance.
+                    // we're going to use it as a hint for getFieldType
+                    bool isExactObj   = false;
+                    bool isNonNullObj = false;
+                    fldOwnerCls       = gtGetClassHandle(base->gtGetOp1(), &isExactObj, &isNonNullObj);
```

Comment:
I wonder how resilient VM code is to mismatched fields and owner types.

Since this is `FIELD_ADDR`, it could be something like:
```
SOME_NODE ref      ; type:       SomeClass
  FIELD_ADDR byref ; owner type: SomeStruct
```
It's hard to get this kind of IR currently, but it depends on a rather fragile fact that you need to access some class field to start using struct fields. If we start optimizing out class fields at zero offset (`m_pEEType`), that'll no longer be the case.

Reply:
> I wonder how resilient VM code is to mismatched fields and owner types.
> 
> Since this is `FIELD_ADDR`, it could be something like:
> 
> ```
> SOME_NODE ref      ; type:       SomeClass
>   FIELD_ADDR byref ; owner type: SomeStruct
> ```
> 
> It's hard to get this kind of IR currently, but it depends on a rather fragile fact that you need to access some class field to start using struct fields. If we start optimizing out class fields at zero offset (`m_pEEType`), that'll no longer be the case.

I am struggling to read your example, did you mean
```
FIELD_ADDR byref ; owner type: SomeStruct
    SOME_NODE ref      ; type:       SomeClass
```
?

Reply:
As an expression it would be: `FIELD_ADDR<SomeStruct::SomeField>(Get<SomeClass>())`.

Reply:
@SingleAccretion I tried to confuse VM here and it was pretty resilient and was throwing MissingFieldException on my attempts. I wrote a unit-test in IL:
```cil
.method static class [System.Runtime]System.Type Test (class MyList`1<string> list)
{
    .maxstack 8

    ldarg.0 // this is MyList`1<string>
    ldfld !0[] class MyList`1<string>::arr
    callvirt instance class [System.Runtime]System.Type [System.Runtime]System.Object::GetType()
    ret
}
```
and was chaing that ldfld type to various types

Reply:
the only thing - I had to insert a "is shared" check for the class

Reply:
(ah, wait, I think I tested a wrong branch)